src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/PackageRestoreInitiator.PackageRestoreInitiatorInstance.cs

```diff
@@ -0,0 +1,240 @@
+                            var versionDropper = CreateVersionDropperBlock();
+                            disposableBag.AddDisposable(sourceBlock.LinkTo(versionDropper, sourceLinkOptions));
+                            return versionDropper.SyncLinkOptions<IProjectValueVersions>(sourceLinkOptions);
+                        });
+
+                    var target = new ActionBlock<Tuple<ImmutableList<IProjectValueVersions>, TIdentityDictionary>>(ProjectPropertyChangedAsync);
+
+                    var targetLinkOptions = new DataflowLinkOptions { PropagateCompletion = true };
+
+                    disposableBag.AddDisposable(ProjectDataSources.SyncLinkTo(sourceBlocks.ToImmutableList(), target, targetLinkOptions));
+
+                    _designTimeBuildSubscriptionLink = disposableBag;
+                }
+            }
+
+            private static IPropagatorBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>, IProjectVersionedValue<IProjectSubscriptionUpdate>> CreateVersionDropperBlock()
+            {
+                var transformBlock = new TransformBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>, IProjectVersionedValue<IProjectSubscriptionUpdate>>(data =>
+                {
+                    return new ProjectVersionedValue<IProjectSubscriptionUpdate>(data.Value, data.DataSourceVersions.RemoveRange(s_keysToDrop));
+                });
+
+                return transformBlock;
+            }
+
+            private Task ProjectPropertyChangedAsync(Tuple<ImmutableList<IProjectValueVersions>, TIdentityDictionary> sources)
+            {
+                IVsProjectRestoreInfo projectRestoreInfo = ProjectRestoreInfoBuilder.Build(sources.Item1, _projectVsServices.Project);
+
+                if (projectRestoreInfo != null)
+                {
+                    _projectVsServices.Project.Services.ProjectAsynchronousTasks
+                        .RegisterAsyncTask(JoinableFactory.RunAsync(async () =>
+                        {
+                            LogProjectRestoreInfo(_projectVsServices.Project.FullPath, projectRestoreInfo);
+
+                            await _solutionRestoreService
+                                   .NominateProjectAsync(_projectVsServices.Project.FullPath, projectRestoreInfo,
+                                        _projectVsServices.Project.Services.ProjectAsynchronousTasks.UnloadCancellationToken)
+                                   .ConfigureAwait(false);
+
+                            CodeMarkers.Instance.CodeMarker(CodeMarkerTimerId.PerfPackageRestoreEnd);
+
+                            CompleteLogProjectRestoreInfo(_projectVsServices.Project.FullPath);
+                        }),
+                        ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
+                        registerFaultHandler: true);
+                }
+
+                return Task.CompletedTask;
```

Comment:
Is there no way to actually await this within this method instead of executing everything synchronously?

Reply:
This is rename, go back to https://github.com/dotnet/project-system/pull/3093 or https://github.com/dotnet/project-system/pull/3097 to review only the code that is added. It might be easier to wait until https://github.com/dotnet/project-system/pull/3093 is merged to see the "real diff".